### PR TITLE
feat(autocomplete): add ability to style md-virtual-repeat-container

### PR DESCRIPTION
**Pull request by annnguyen**
…er generated by `md-autocomplete`. 

This allows control over css styling for specific instances of `<md-virtual-repeat-container>` created by `md-autocomplete`.

Fixes [#8393](https://github.com/angular/material/issues/8393)

### CODE REVIEW

**Comment by annnguyen** [comment]
I signed it!

**Comment by BevanR** [comment]
Thanks for your contribution.  I responded on #8393 with some ideas to make this easier to review.

**Comment by BevanR** [comment]
A quick look over the code looks good.  It has a test, which is great!

I think it would be more useful if it allowed class(es) to be set instead of IDs.  IDs must be unique in a document.  However classes can be re-used.  If only IDs can be set, then a component producing many `<md-autocomplete>` elements (such as a directive that expands some element into multiple `<md-autocomplete>` elements) will need to ensure each one has unique ID, and the consumer's CSS will not be able to efficiently modify any elements coming from the factory-component.

**Comment by BevanR** [comment]
This appears to duplicate #8403.  I like the approach in #8403 better because it allows classes to be set, instead of IDs.  However this PR has tests.  Perhaps @vassagus and you can merge efforts?

**Comment by annnguyen** [comment]
Thank @BevanR. I'll make updates to the ticket and pull request code. @vassagus if you're okay, i can update my code to set class to md-virtual-repeat-container instead of id if that's the preferred method. Lmk!

**Comment by vassagus** [comment]
@annnguyen sounds good, you can go ahead and merge my commit. Thanks guys.

**Comment by annnguyen** [comment]
@DevVersion code is ready for review whenever you guys have bandwidth!

**Comment by ThomasBurleson** [comment]
This issue is closed as part of our ‘Surge Focus on Material 2' efforts.
      For details, see our forum posting @ http://bit.ly/1UhZyWs.

**Comment by ausminternet** [comment]
:rage: I can't count anymore the bugs I need to be fixed and which are closed in favour of a not-yet-released version 2... Even if there are PRs...

**Comment by jensbodal** [comment]
Using information from this pull request I made a decorator to add to my app's config;  the decorator allows an `md-menu-container-class` attribute to be added to the `md-autocomplete` directive which can then be used for styling.  [Plunker](https://plnkr.co/edit/0tzee8RpoD2WeFSJX1z1?p=preview)

**Comment by ThomasBurleson** [comment]
@jensbodal - This is a nice technique to extend/override an existing ngM1 component DDO!  :-). Perhaps you should blog this for the community ;-).

>

**Comment by vassagus** [comment]
@jensbodal thank you, that works for me as well.

**Comment by dohomi** [comment]
@jensbodal I think this is exactly what the current `md-autocomplete` should do. Would be great if this issue would be opened to add the class to the outer container.

**Comment by jensbodal** [comment]
@ThomasBurleson thanks for the comment and the suggestion.  It took me a bit to decide on a proper blog framework but decided to use codepen as it seems to fit.  Blog post made available here: http://codepen.io/jensbodal/post/extending-an-angular-ddo-directive-definition-object

**Comment by tiagomsmagalhaes** [comment]
@jensbodal thanks for the decorator. I had to tweak it because i had `md-autocomplete` in another directive so i had to pass it as a variable, so I tweaked it to allow that. Is there a way that could allow both options so we could merge it? my codepen: http://codepen.io/tiagomagalhaes/pen/dOXPXj

**Comment by jensbodal** [comment]
@tiagomsmagalhaes you can just check to see if the element resolves to include an expression, and if not read from the attribute as a string directly.

http://codepen.io/jensbodal/pen/oYLWxr

*(relevant part)*
```js
return function(e, a) {
  var menuContainer = a.find('md-virtual-repeat-container');
  var menuContainerClass = '';
  
  if (typeof e.menuContainerClass !== 'boolean' && e.menuContainerClass !== 0) {
    menuContainerClass = e.menuContainerClass;      
  }

  else {
    menuContainerClass = attr.mdMenuContainerClass || '';
  }
  
  menuContainer.addClass(menuContainerClass);

  template.apply(this, arguments);
};
```

I'm looking for `!== 'boolean' && !== 0` as that seemed to fit all the cases where the attribute didn't exist (`false`), the attribute referenced a controller variable which didn't exist (`true`), the attribute was a string and contained dashes (`0`), and the attribute was a string (`true`).  The `!== 0` is related to how angular interpolates attributes I believe.

File: src/components/autocomplete/js/autocompleteDirective.js [hunk 1]
```diff
@@ -181,13 +184,15 @@ function MdAutocomplete () {
...
+              id="{{menuContainerId}}"\
               md-auto-shrink\
               md-auto-shrink-min="1"\
               ng-mouseenter="$mdAutocompleteCtrl.listEnter()"\
               ng-mouseleave="$mdAutocompleteCtrl.listLeave()"\
               ng-mouseup="$mdAutocompleteCtrl.mouseUp()"\
               ng-hide="$mdAutocompleteCtrl.hidden"\
-              class="md-autocomplete-suggestions-container md-whiteframe-z1"\
+              class="md-autocomplete-suggestions-container md-whiteframe-z1' +
+                (attr.mdMenuContainerClass != null ? ' ' + attr.mdMenuContainerClass : '') + '"\
```
Comment (Splaktar) [inline src/components/autocomplete/js/autocompleteDirective.js]: Can this use `ng-class` like `menuClass` uses below?

Comment (Splaktar) [reply src/components/autocomplete/js/autocompleteDirective.js]: Looks like that isn't needed, but the logic here can be pulled out of the template to make things cleaner.

File: src/components/autocomplete/js/autocompleteDirective.js [hunk 2]
```diff
@@ -181,13 +184,15 @@ function MdAutocomplete () {
               ng-if="$mdAutocompleteCtrl.loadingIsVisible()"\
               md-mode="indeterminate"></md-progress-linear>\
           <md-virtual-repeat-container\
+              id="{{menuContainerId}}"\
```
Comment (Splaktar) [inline src/components/autocomplete/js/autocompleteDirective.js]: This looks to be left over from the previous implementation and should be removed.

**Comment by Splaktar** [comment]
Moved this over to a new PR https://github.com/angular/material/pull/11194

**Pull request closed without merging**